### PR TITLE
fix: error type hints in Python3.12 (#1147)

### PR DESCRIPTION
**Pull request by 954-Ivory**
Issue: https://github.com/docarray/docarray/issues/1839

---
### Description

I encountered the following issue when using langchain in Python 3.12:

https://github.com/langchain-ai/langchain/issues/15016

```console
    return Generic.__class_getitem__.__func__(cls, item)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'builtin_function_or_method' object has no attribute '__func__'. Did you mean: '__doc__'?
```

It caused by this PR (https://github.com/docarray/docarray/pull/1147)

### Example Code

```Python
https://learn.deeplearning.ai/langchain/lesson/5/question-and-answer
```

### CODE REVIEW

**Review by samsja — CHANGES_REQUESTED** [review]
Thank you very much for the PR. 

I am not sure about the implementation. I think it would be nicer if do a if else on the python version rather that relying on catching the exception

**Comment by JoanFM** [comment]
Hey @954-Ivory , we need you to sign off the commits and to comply with our guidelines as per lur CONTRIBUTING_GUIDELINES

File: docarray/array/any_array.py [hunk 1]
```diff
@@ -51,8 +51,13 @@ def __repr__(self):
...
         if not isinstance(item, type):
-            return Generic.__class_getitem__.__func__(cls, item)  # type: ignore
-            # this do nothing that checking that item is valid type var or str
+            try:
+                return Generic.__class_getitem__.__func__(cls, item)  # type: ignore
+                # this do nothing that checking that item is valid type var or str
+                # Keep the approach in #1147 to be compatible with lower versions of Python.
+            except AttributeError:
+                from types import GenericAlias
+                return GenericAlias(cls, item)  # type: ignore
```
Comment (samsja) [inline docarray/array/any_array.py:60]:
```suggestion
            python_minor_version = sys.version_info.minor
            if python_minor_version < 12:
                return Generic.__class_getitem__.__func__(cls, item)  # type: ignore
                # this do nothing that checking that item is valid type var or str
                # Keep the approach in #1147 to be compatible with lower versions of Python.
            else:
                 from types import GenericAlias
                 return GenericAlias(cls, item)  # type: ignore
```

Comment (954-Ivory) [reply docarray/array/any_array.py:60]: > Maybe ```sys.version_info < (3, 12)``` is better.

Comment (samsja) [reply docarray/array/any_array.py:60]: yes true

Comment (JoanFM) [reply docarray/array/any_array.py]: let's keep the computation of the boolean on the top of the file.

**Review by samsja — APPROVED** [review]

**Comment by samsja** [comment]
I approved, but before merging can you apply black @954-Ivory ? and DCO as well ( you can do `git rebase HEAD~2 --signoff` and the `git push -f`

**Comment by 954-Ivory** [comment]
> I approved, but before merging can you apply black @954-Ivory ? and DCO as well ( you can do `git rebase HEAD~2 --signoff` and the `git push -f`

Can you help me check if I have made any mistakes?

**Pull request merged**